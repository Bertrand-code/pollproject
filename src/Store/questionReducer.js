import { fetchLeaderBoardStat } from "../db";
import { _saveQuestion, _saveQuestionAnswer } from "../db/_DATA";

const initialState = {
  questions: [],
  answers: {},
  leaderboard: null,
};

export default function questionReducer(
  state = initialState,
  { type, payload }
) {
  switch (type) {
    case "question_loaded":
      return {
        ...state,
        questions: [...payload.questions],
        answers: { ...state.answers, ...payload.answers },
      };
    case "answer_saved":
      return {
        ...state,
        answers: { ...state.answers, [payload.qid]: payload.answer },
      };
    case "question_saved":
      return {
        ...state,
        questions: [...state.questions, payload],
      };
    case "leaderboard_loaded":
      return {
        ...state,
        leaderboard: payload,
      };

    case "logout":
      return { questions: [], answers: {}, leaderboard: null };

    default:
      return state;
  }
}

export const loadQuestionAction = (questions, answers) => {
  return {
    type: "question_loaded",
    payload: {
      questions,
      answers,
    },
  };
};

export const saveAnswerAction = (qid, answer) => {
  return {
    type: "answer_saved",
    payload: {
      qid,
      answer,
    },
  };
};
const loadLeaderboardAction = (leaderboard) => {
  return {
    type: "leaderboard_loaded",
    payload: leaderboard,
  };
};

const questionSaved = (payload) => {
  return {
    type: "question_saved",
    payload,
  };
};

export const saveQuestionAction = (question) => {
  return async (dispatch, getState) => {
    const { id: userId } = getState().authStore.user;
    const { id } = await _saveQuestion({
      author: userId,
      optionOneText: question.optionOne,
      optionTwoText: question.optionTwo,
    });
    dispatch(questionSaved(id));
  };
};

export const answerQuestion = (qid, answer) => {
  return async (dispatch, getState) => {
    const userId = getState().authStore.user.id;
    _saveQuestionAnswer({ authedUser: userId, qid, answer }).then((res) => {
      if (res) dispatch(saveAnswerAction(qid, answer));
    });
  };
};

export const loadLeaderboard = () => {
  return async (dispatch) => {
    const leaderboard = await fetchLeaderBoardStat();
    dispatch(loadLeaderboardAction(leaderboard));
  };
};
