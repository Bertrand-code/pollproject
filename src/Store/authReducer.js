import { login } from "../db";
import { loadQuestionAction } from "./questionReducer";

const initialValues = {
  user: null,
  auth: false,
  loading: false,
  error: null,
};

export default function authReducer(state = initialValues, { type, payload }) {
  switch (type) {
    case "user_logged":
      return { ...state, auth: true, user: payload };
    case "loading":
      return { ...state, loading: !state.loading };
    case "error":
      return { ...state, error: payload };
    case "logout":
      return { ...state, auth: false, user: null };
    default:
      return state;
  }
}

const loginAction = (payload) => ({ type: "user_logged", payload });
const loadingAction = () => ({ type: "loading" });
export const logoutAction = () => ({ type: "logout" });
export const userLogin = (userInfo) => async (dispatch) => {
  dispatch(loadingAction());
  try {
    const { answers, questions, ...rest } = await login(
      userInfo.username,
      userInfo.password
    );
    dispatch(loginAction(rest));
    dispatch(loadQuestionAction(questions, answers));
  } catch (error) {
    dispatch(loadingAction());
    dispatch({ type: "error", payload: error.message });
  }
};
