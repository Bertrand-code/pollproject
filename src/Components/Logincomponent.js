import React, { useState } from "react";
import { connect } from "react-redux";
import { users } from "../db/_DATA";
import { Navigate, useLocation } from "react-router-dom";
import { userLogin } from "../Store/authReducer";

function LoginComponent({ login, auth }) {
  const [current, setCurrent] = useState(users.sarahedo);

  const handleSubmit = (e) => {
    e.preventDefault();
    login({ username: current.id, password: current.password });
  };

  let location = useLocation();
  let from = location.state?.from?.pathname || "/questions";

  const onchange = (e) => {
    setCurrent(users[e.target.value]);
  };

  if (auth) return <Navigate to={from} replace />;
  return (
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        width: "100%",
        height: "100vh",
      }}
    >
      <form
        className="loginForm"
        onSubmit={handleSubmit}
        data-testid="loginForm"
      >
        <label>
          Username:
          <select name="username" onChange={onchange}>
            {Object.values(users).map((user) => (
              <option value={user.id} key={user.id}>
                {user.id}
              </option>
            ))}
          </select>
          <br />
        </label>
        <br />
        <label>
          Password:
          <input
            className="inputElement"
            type="password"
            value={current.password}
            onChange={(e) => {}}
          />
        </label>
        <br />
        <button className="btn"> Login </button>
      </form>
    </div>
  );
}

const mapDispatchToProps = {
  login: userLogin,
};
const mapStateToProps = (state) => ({
  auth: state.authStore.auth,
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginComponent);
