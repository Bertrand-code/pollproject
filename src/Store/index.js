import {
  applyMiddleware,
  combineReducers,
  compose,
  legacy_createStore as createStore,
} from "redux";
import thunk from "redux-thunk";
import authReducer from "./authReducer";
import questionReducer from "./questionReducer";

export const reducer = combineReducers({
  authStore: authReducer,
  questionsStore: questionReducer,
});

export const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk)
    // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  )
);
