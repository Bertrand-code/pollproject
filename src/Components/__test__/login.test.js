import { render, cleanup, screen } from "@testing-library/react";
import Logincomponent from "../Logincomponent";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../../Store";

afterEach(() => {
  cleanup();
});

const TestLogin = () => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <Routes>
          <Route path="/" element={<Logincomponent />} />
        </Routes>
      </Provider>
    </BrowserRouter>
  );
};

describe("Logincomponent", () => {
  it("should render and match screenshot", () => {
    const { container } = render(<TestLogin />);
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should render form", () => {
    render(<TestLogin />);
    expect(screen.getByTestId("loginForm")).toBeInTheDocument();
  });
  it("should render login button", () => {
    render(<TestLogin />);
    expect(screen.getByText("Login")).toBeInTheDocument();
  });
});
