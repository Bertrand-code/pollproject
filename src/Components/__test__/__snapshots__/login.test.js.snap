// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logincomponent should render and match screenshot 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center; width: 100%; height: 100vh;"
  >
    <form
      class="loginForm"
      data-testid="loginForm"
    >
      <label>
        Username:
        <select
          name="username"
        >
          <option
            value="sarahedo"
          >
            sarahedo
          </option>
          <option
            value="tylermcginnis"
          >
            tylermcginnis
          </option>
          <option
            value="mtsamis"
          >
            mtsamis
          </option>
          <option
            value="zoshikanlu"
          >
            zoshikanlu
          </option>
        </select>
        <br />
      </label>
      <br />
      <label>
        Password:
        <input
          class="inputElement"
          type="password"
          value="password123"
        />
      </label>
      <br />
      <button
        class="btn"
      >
         Login 
      </button>
    </form>
  </div>
</div>
`;
